### PR TITLE
Add the Picard task AddOrReplaceReadGroups

### PR DESCRIPTION
**Pull request by clintval**

### CODE REVIEW

File: tasks/src/main/scala/dagr/tasks/picard/AddOrReplaceReadGroups.scala [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  val platformUnit: String,
+  val sampleName: String,
+  val sequencingCenter: Option[String] = None,
+  val description: Option[String] = None,
+  val runDate: Option[Iso8601Date] = None,
+  val keySequence: Option[String] = None,
+  val readGroupFlowOrder: Option[String] = None,
+  val predictedInsertSize: Option[Int] = None,
+  val programGroup: Option[String] = None,
+  val platformModel: Option[String] = None
```
Comment (clintval) [inline tasks/src/main/scala/dagr/tasks/picard/AddOrReplaceReadGroups.scala:48]: The fields are ordered as they are on the Picard CLI which puts two optionals (`sortOrder` and `id`) before required fields.

Would you like the constructor fields sorted "required -> optional"?

File: tasks/src/main/scala/dagr/tasks/picard/AddOrReplaceReadGroups.scala [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import dagr.tasks.DagrDef.PathToBam
+import htsjdk.samtools.SAMFileHeader.SortOrder
+import htsjdk.samtools.util.Iso8601Date
+
+import scala.collection.mutable.ListBuffer
+
+object AddOrReplaceReadGroups {
+
+  /** The Picard default for the Library ID tag. */
+  val DefaultLibraryId: String = "1"
```
Comment (clintval) [inline tasks/src/main/scala/dagr/tasks/picard/AddOrReplaceReadGroups.scala]: https://github.com/broadinstitute/picard/blob/b02e42e0ca82efae70ee572766cbe101b6a2305e/src/main/java/picard/sam/AddOrReplaceReadGroups.java#L105-L106

File: tasks/src/main/scala/dagr/tasks/picard/AddOrReplaceReadGroups.scala (hunk 1 above)
Comment (clintval) [reply tasks/src/main/scala/dagr/tasks/picard/AddOrReplaceReadGroups.scala:48]: I just re-ordered "require -> optional". The API is better IMHO.

File: tasks/src/main/scala/dagr/tasks/picard/AddOrReplaceReadGroups.scala [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+import dagr.tasks.DagrDef.PathToBam
+import htsjdk.samtools.SAMFileHeader.SortOrder
+import htsjdk.samtools.util.Iso8601Date
+
+import scala.collection.mutable.ListBuffer
+
+object AddOrReplaceReadGroups {
+
+  /** The Picard default for the Library ID tag. */
```
Comment (nh13) [inline tasks/src/main/scala/dagr/tasks/picard/AddOrReplaceReadGroups.scala]:
```suggestion
  /** The default for the Library ID tag in Picard's `AddOrReplaceReadGroups` tool. */
```

File: tasks/src/main/scala/dagr/tasks/picard/AddOrReplaceReadGroups.scala (hunk 2 above)
Comment (nh13) [reply tasks/src/main/scala/dagr/tasks/picard/AddOrReplaceReadGroups.scala]: serious

File: tasks/src/main/scala/dagr/tasks/picard/AddOrReplaceReadGroups.scala [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+class AddOrReplaceReadGroups(
+  val in: PathToBam,
+  val out: PathToBam,
+  val sampleName: String,
+  val library: String,
+  val platform: String,
+  val platformUnit: String,
+  val id: Option[String] = Some(AddOrReplaceReadGroups.DefaultLibraryId),
```
Comment (nh13) [inline tasks/src/main/scala/dagr/tasks/picard/AddOrReplaceReadGroups.scala]: can't this just be `None`, and then we can rely upon Picard to set the default?

File: tasks/src/main/scala/dagr/tasks/picard/AddOrReplaceReadGroups.scala [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  val programGroup: Option[String] = None,
+  val platformModel: Option[String] = None
+) extends PicardTask {
+
+  override protected def addPicardArgs(buffer: ListBuffer[Any]): Unit = {
+    buffer.append("I=" + in)
+    buffer.append("O=" + out)
+    buffer.append("RGSM=" + sampleName)
+    buffer.append("RGLB=" + library)
+    buffer.append("RGPL=" + platform)
```
Comment (nh13) [inline tasks/src/main/scala/dagr/tasks/picard/AddOrReplaceReadGroups.scala]: Don't change this, but as a note that `platform` has only a fixed set of spec-compliant values.

**Review by nh13 — CHANGES_REQUESTED** [review]
One minor nit about the default value for `id`.  Thanks for adding this.

File: tasks/src/main/scala/dagr/tasks/picard/AddOrReplaceReadGroups.scala (hunk 2 above)
Comment (clintval) [reply tasks/src/main/scala/dagr/tasks/picard/AddOrReplaceReadGroups.scala]: I simply removed, I actually don't understand my original motivation anymore.

**Pull request merged**